src/object_changeset.cpp

```diff
@@ -0,0 +1,156 @@
+        }
+    }
+}
+
+void ObjectChangeSet::deletions_add(ObjectKeyType obj)
+{
+    m_modifications.erase(obj);
+    size_t num_inserts_removed = m_insertions.erase(obj);
+    if (num_inserts_removed == 0) {
+        m_deletions.insert(obj);
+    }
+}
+
+void ObjectChangeSet::clear(size_t old_size)
+{
+    static_cast<void>(old_size); // unused
+    m_clear_did_occur = true;
+    m_insertions.clear();
+    m_modifications.clear();
+    m_deletions.clear();
+}
+
+bool ObjectChangeSet::insertions_remove(ObjectKeyType obj)
+{
+    return m_insertions.erase(obj) > 0;
+}
+
+bool ObjectChangeSet::modifications_remove(ObjectKeyType obj)
+{
+    return m_modifications.erase(obj) > 0;
+}
+
+bool ObjectChangeSet::deletions_remove(ObjectKeyType obj)
+{
+    return m_deletions.erase(obj) > 0;
+}
+
+bool ObjectChangeSet::deletions_contains(ObjectKeyType obj) const
+{
+    if (m_clear_did_occur) {
+        // FIXME: what are the expected notifications when an object is deleted
+        // and then another object is inserted with the same key?
+        return m_insertions.count(obj) == 0;
+    }
+    return m_deletions.count(obj) > 0;
+}
+
+bool ObjectChangeSet::insertions_contains(ObjectKeyType obj) const
+{
+    return m_insertions.count(obj) > 0;
+}
+
+bool ObjectChangeSet::modifications_contains(ObjectKeyType obj) const
+{
+    return m_modifications.count(obj) > 0;
+}
+
+const ObjectChangeSet::ObjectSet* ObjectChangeSet::get_columns_modified(ObjectKeyType obj) const
+{
+    auto it = m_modifications.find(obj);
+    if (it == m_modifications.end()) {
+        return nullptr;
+    }
+    return &it->second;
+}
+
+void ObjectChangeSet::merge(ObjectChangeSet&& other)
+{
+    if (other.empty())
+        return;
+    if (empty()) {
+        *this = std::move(other);
+        return;
+    }
+    m_clear_did_occur = m_clear_did_occur || other.m_clear_did_occur;
+
+    verify();
+    other.verify();
+
+    // Drop any inserted-then-deleted rows, then merge in new insertions
+    for (auto it = other.m_deletions.begin(); it != other.m_deletions.end();) {
+        auto previously_inserted = m_insertions.find(*it);
+        auto previously_modified = m_modifications.find(*it);
+        if (previously_modified != m_modifications.end()) {
+            m_modifications.erase(previously_modified);
+        }
+        if (previously_inserted != m_insertions.end()) {
+            m_insertions.erase(previously_inserted);
+            it = m_deletions.erase(it);
+        }
+        else {
+            ++it;
+        }
+    }
+    if (!other.m_insertions.empty()) {
+        m_insertions.insert(other.m_insertions.begin(), other.m_insertions.end());
+    }
+    if (!other.m_deletions.empty()) {
+        m_deletions.insert(other.m_deletions.begin(), other.m_deletions.end());
+    }
+    for (auto it = other.m_modifications.begin(); it != other.m_modifications.end(); ++it) {
+        auto insert_result = m_modifications.insert(*it);
```

Comment:
This can also be `m_modifications[it->first].insert(it->second.begin(), it->second.end());`